tensorizer/stream_io.py

```diff
@@ -162,17 +174,62 @@ def __init__(
 
         self._curl = None
         with _wide_pipes.widen_new_pipes():  # Widen on Windows
+            popen_start = time.time()
             self._curl = subprocess.Popen(
                 cmd,
                 stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
                 bufsize=256 * 1024 * 1024,
             )
+        popen_end = time.time()
 
         _wide_pipes.widen_pipe(self._curl.stdout.fileno())  # Widen on Linux
+        resp = self._curl.stderr.readline() # Block on the http header response
+        resp_begin = time.time()
+
+        # We reinitialize this object when seeking, so we don't want to overwrite
+        # these tracking variables if they already exist.
+        self._init_vars()
+
+        # Track the latency of the popen and http response
+        self.popen_latencies.append(popen_end - popen_start)
+        self.http_response_latencies.append(resp_begin - popen_end)
+
+        if not resp.startswith(b"HTTP/1.1 2") and not resp.startswith(b"HTTP/2 2"):
+            self._curl.terminate()
+            self._curl.wait()
+            raise IOError(f"Failed to open stream: {resp.decode('utf-8')}")
+        # Make stderr non-blocking so that we can read the rest of the header without
+        # blocking the main thread.
+        os.set_blocking(self._curl.stderr.fileno(), False)
+        # Read the rest of the header response and parse it.
+        header_lines = self._curl.stderr.readlines()
+        self.response_headers = {}
+        for line in header_lines:
+            line = line.decode("utf-8").strip()
+            if not line or ':' not in line:
+                continue
+            k, v = line.split(":", maxsplit=1)
+            self.response_headers[k.strip()] = v.strip()
```

Comment:
two subtle bug I just caught here:
1. multiple headers with the same name are *valid* in HTTP, and have special concatenation rules. [most headers should be concatenated with commas, but there are exceptions].
2. the header names are not case sensitive.

maybe a good enough solution is to just alwatys concat here, with commas.

Reply:
Good catch. I'd say that starts falling into the realm of 'just use a HTTP parsing lib, dummy'. I'll add a` FIXME` note.

Reply:
`FIXME` added, and lowercased header keys.

Reply:
Fixed this by using Python's built-in `http.client.parse_headers` routine in 558176610d5b159b7bb7cb97304992f99e511b4f